---
title: "Simulations from 10x-genomic dataset, fold change 2"
author: "Fanny Perraudeau & Koen Van den Berge"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_height: 7
    fig_width: 7
    toc: yes
    code_folding: hide
    toc_float: yes
---

```{r options, echo=FALSE, results="hide",message=FALSE, error=FALSE, include=FALSE, autodep=TRUE, warning=FALSE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(zinbwave)
library(BiocParallel)
library(doParallel)
library(Biobase)
library(edgeR)
library(scales)
library(DESeq2)
library(iCOBRA) # roc
library(limma)
library(genefilter) #filtered pvalues
library(MAST)
library(RColorBrewer)
library(knitr)
library(ggplot2)
library(cowplot)
# use new simulation.
pathToParentFolder="~/Dropbox/phdKoen/singleCell/zinbwavezingerGitHub/zinbwaveZinger/"
source(paste0(pathToParentFolder,"zingeRsimulationFunctions/simulationHelpFunctions_v7_diffInZero.R"))
```

```{r}
set.seed(8283) #just to make sure rmd is reproducible
```

This document was written by Fanny and adapted by Koen for the novel simulation framework.

The goal of this document is to evaluate the performance on 10X Genomics scRNA-seq data. We simulate data from a 10X-genomics dataset using the updated simulation framework from zingeR team.

# Simulate scRNA-seq data

## Real dataset

```{r data}
install_cellranger = FALSE
if (install_cellranger){
  source("http://cf.10xgenomics.com/supp/cell-exp/rkit-install-2.0.0.R")
  library(cellrangerRkit)
  packageVersion("cellrangerRkit")
}
library(cellrangerRkit)
pipestance_path <- "./"
if(!file.exists(paste0(pipestance_path, "/outs"))) {
  download_sample(sample_name="pbmc4k",sample_dir=pipestance_path,
                  host="http://cf.10xgenomics.com/samples/cell-exp/1.3.0/")
}
pbmc <- load_cellranger_matrix(pipestance_path)
analysis_results <- load_cellranger_analysis_results(pipestance_path)
dim(pbmc)
```

```{r}
counts = exprs(pbmc) # expression matrix
counts[1:5,1:5]
head(fData(pbmc)) # data frame of genes
head(pData(pbmc)) # data frame of cell barcodes
```

In the 10X website, the authors performed "graph-based clustering". We will use those labels here.

```{r}
cl <- analysis_results[["clustering"]][["graphclust"]][,'Cluster']
names(cl) <- analysis_results[["clustering"]][["graphclust"]][,'Barcode']
cl <- cl[colnames(pbmc)]
```

```{r}
pal <- clusterExperiment::bigPalette
plot(analysis_results[['tsne']][,2:3], col = pal[cl],
     xlim = c(-40, 60), ylim = c(-60, 40))
```

```{r}
df = data.frame(tsne1 = analysis_results[['tsne']][,2],
                tsne2 = analysis_results[['tsne']][,3],
                Cluster=as.factor(cl))
df$CD3D <- log1p(exprs(pbmc)[which(fData(pbmc)[,2] == "CD3D"),]) #t cell
df$IL7R <- log1p(exprs(pbmc)[which(fData(pbmc)[,2] == "IL7R"),])
df$NKG7 <- log1p(exprs(pbmc)[which(fData(pbmc)[,2] == "NKG7"),]) #nk cells
df$S100A8 <- log1p(exprs(pbmc)[which(fData(pbmc)[,2] == "S100A8"),]) #myeloid cells
df$CCR10 <- log1p(exprs(pbmc)[which(fData(pbmc)[,2] == "CCR10"),]) #mem t cells
df$TNFRSF18 <- log1p(exprs(pbmc)[which(fData(pbmc)[,2] == "TNFRSF18"),]) #reg t cells
df$ID3 <- log1p(exprs(pbmc)[which(fData(pbmc)[,2] == "ID3"),]) #naive t cells
df$PF4 <- log1p(exprs(pbmc)[which(fData(pbmc)[,2] == "PF4"),]) #megakaryocytes

ggplot(aes(tsne1, tsne2, colour=CD3D), data = df) +
  geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(tsne1, tsne2, colour=NKG7), data = df) +
  geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(tsne1, tsne2, colour=S100A8), data = df) +
  geom_point() + scale_colour_gradient(low="blue", high="yellow")
```

Let's subset the data to only two cell types.
```{r}
# red = myeloid cells
# green = ?
keep = cl %in% c(2,3)
plot(analysis_results[['tsne']][,2:3][keep, ], col = pal[cl][keep],
     xlim = c(-40, 60), ylim = c(-60, 40))
```

```{r}
counts = counts[, cl %in% 2:3]
cellType = cl[cl %in% 2:3]
```

Input count matrix for zinbwave should not have genes with only zero counts. Let's remove these genes right away.

```{r}
counts <- counts[rowSums(counts != 0) > 20, ] #expression in at least 1/60 cells
dim(counts)
tenXData = counts
```


## simulation framework

```{r}
paramsTenx = getDatasetMoMPositive(counts = tenXData)
```

```{r sims}
nSamples=ncol(tenXData)
grp=as.factor(rep(0:1, each = nSamples/2)) #two-group comparison
nTags=10000 #nr of features
set.seed(11)
DEind = sample(1:nTags,floor(nTags*.1),replace=FALSE) #10% DE
fcSim=(2 + rexp(length(DEind), rate = 1/2)) #fold changes
libSizes=sample(colSums(tenXData),nSamples,replace=TRUE) #library sizes
simData <- NBsimSingleCell(foldDiff = fcSim, ind = DEind,
                           dataset = tenXData, nTags = nTags,
                           group = grp,
                           verbose = TRUE, params = paramsTenx,
                           lib.size = libSizes, cpm="AveLogCPM", normalizeLambda=TRUE, min.dispersion=1e-3)
simData$counts[1:5,1:5]
```

```{r}
# BCV plots
dOrig=suppressWarnings(edgeR::calcNormFactors(DGEList(tenXData)))
dOrig = estimateDisp(dOrig, design=model.matrix(~cellType))

d=suppressWarnings(edgeR::calcNormFactors(DGEList(simData$counts)))
d=estimateDisp(d, design=model.matrix(~grp))
```

```{r}
par(mfrow=c(1,2))
plotBCV(dOrig,ylim=c(0,12), xlim=c(9,16))
plotBCV(d,ylim=c(0,12), xlim=c(9,16))
par(mfrow=c(1,1))

```

```{r}
# association of library size with zeros
plot(x=log(colSums(tenXData)), y=colMeans(tenXData==0), xlab="Log library size", ylab="Fraction of zeros", xlim=c(5.5,20))
points(x=log(colSums(simData$counts)), y=colMeans(simData$counts==0), col=2)
```

```{r}
# association of aveLogCPM with zeros
plot(x=edgeR::aveLogCPM(tenXData), y=rowMeans(tenXData==0), xlab="Average log CPM", ylab="Fraction of zeros", ylim=c(0,1), col=alpha(1,1/2), pch=19, cex=.3)
points(x=edgeR::aveLogCPM(simData$counts),
       y=rowMeans(simData$counts==0),col=alpha(2,1/2),pch=19,cex=.3)
```

# Methods
## RNA-seq methods
### edgeR
```{r}
edgeR <- function(counts, group, ylim = NULL, xlim = NULL){
  d <- DGEList(counts)
  d <- suppressWarnings(edgeR::calcNormFactors(d))
  design <- model.matrix(~group)
  d <- estimateDisp(d, design)
  plotBCV(d, ylim = ylim, main = 'edgeR', xlim = xlim)
  fit <- glmFit(d,design)
  lrt <- glmLRT(fit, coef = 2)
  pval <- lrt$table$PValue
  padj <- p.adjust(pval, "BH")
  cbind(pval = pval, padj = padj)
}
```

### DESeq2
```{r}
DESeq2 <- function(counts, group, ylim = NULL, xlim = NULL){
  colData <- data.frame(group = group)
  dse <- DESeqDataSetFromMatrix(countData=counts, colData=colData, design=~group)
  colData(dse)$group <- as.factor(colData(dse)$group)
  dse <- estimateSizeFactors(dse, type="poscounts")
  dse <- estimateDispersions(dse, minmu=1e-3)
  dse = nbinomLRT(dse, minmu=1e-3, reduced=~1)
  rr <- results(dse)
  cbind(pval = rr$pvalue, padj = rr$padj)
}
```

### limma-voom
```{r}
limma <- function(counts, group, ylim = NULL, xlim = NULL){
	design <- model.matrix(~ group)
	nf <- suppressWarnings(edgeR::calcNormFactors(counts))
	y <- voom(counts, design, plot = FALSE, lib.size = colSums(counts) * nf)
	fit <- lmFit(y, design)
	fit <- eBayes(fit)
	tt <- topTable(fit, coef = 2, n = nrow(counts), sort.by = "none")
	pval <- tt$P.Value
	padj <- tt$adj.P.Val
	cbind(pval = pval, padj = padj)
}
```

## scRNA-seq methods
We encounter errors with the latest version of scde, as documented here: https://groups.google.com/forum/#!topic/singlecellstats/rbFUTOQ9wu4. We followed the guidelines suggested by the authors and work with version 1.99.2.

```{r scde}
scde <- function(counts, group, ylim=NULL, xlim=NULL){
  counts = matrix(as.integer(counts),nrow=nrow(counts),ncol=ncol(counts))
  if(is.null(colnames(counts))) colnames(counts)=paste0("sample",1:ncol(counts))
  require(scde)

  # calculate error models
  o.ifm <- scde.error.models(counts = counts, groups = group, n.cores = 1, threshold.segmentation = TRUE, save.crossfit.plots = FALSE, save.model.plots = FALSE, verbose = 0)
  # estimate gene expression prior
  o.prior <- scde.expression.prior(models = o.ifm, counts = counts, length.out = 400, show.plot = FALSE)
  # calculate differential expression
  ediff <- scde.expression.difference(o.ifm, counts, o.prior, groups  =  group, n.randomizations  =  150, n.cores  =  1, verbose  =  0)
  lfc <- ediff$mle
  pval=(1-pnorm(abs(ediff$Z)))*2
  padj=(1-pnorm(abs(ediff$cZ)))*2
  out = cbind(pval,padj,lfc)
  return(out)
}

```

### MAST

```{r MAST}
### copied code from FPR_mocks.Rmd on September 14, 2017.
MAST <- function(counts, group, ylim = NULL, xlim = NULL){
  require(MAST)
  tpm <- counts*1e6/colSums(counts)
  tpm <- log2(tpm+1)
  sca <- FromMatrix(tpm,  cData=data.frame(group=group))
  #sca <- FromMatrix(counts,  cData=data.frame(group=group))

  # Adaptive thresholding from MAST vignette
  freq_expressed <- 0.05
  thres <- thresholdSCRNACountMatrix(assay(sca), nbins = 300, min_per_bin = 50, conditions = group)
  assays(sca) <- list(thresh=thres$counts_threshold, tpm=assay(sca))
  expressed_genes <- freq(sca) > freq_expressed
  sca <- sca[expressed_genes,]

  ngeneson <- apply(counts,2,function(x) mean(x>0))
  CD <- colData(sca)
  CD$ngeneson <- ngeneson
  CD$cngeneson <- CD$ngeneson-mean(ngeneson)
  colData(sca) <- CD
  ## differential expression
  fit <- zlm(~ cngeneson + group , sca = sca)
  lrFit <- lrTest(fit, 'group')
  pval <- lrFit[, 'hurdle', 'Pr(>Chisq)']
  padj <- p.adjust(pval, method = "BH")

  ### MAST filtered the genes, so make a list that is consistent with the original count matrix.
  pvalAll = vector(length=nrow(counts))
  pvalAll[] = 1
  names(pvalAll)=rownames(counts)
  pvalAll[match(names(pval),names(pvalAll))] = pval

  padjAll = vector(length=nrow(counts))
  padjAll[] = 1
  names(padjAll)=rownames(counts)
  padjAll[match(names(padj),names(padjAll))] = padj

  out = cbind(pval = pvalAll, padj = padjAll, logfc = NA)
  return(out)
}

```

### NODES

```{r}
NODES <- function(counts, group, xlim, ylim){
  require(NODES)
  g=ifelse(group==0,"A","B")
  colnames(counts)=g
  normCounts=pQ(counts)
  res=NODES::NODES(data=normCounts,group=colnames(normCounts))
  pval=vector(length=nrow(counts))
  names(pval)=rownames(counts)
  pval[rownames(normCounts)]=res$Fisher
  pval[is.na(pval)]=1
  padj=p.adjust(pval,"BH")
  lfc=NA
  out=cbind(pval,padj,lfc)
  return(out)
}
```

### metagenomeSeq

```{r metagenomeSeq}
metagenomeSeq <- function(counts, group, xlim, ylim){
  require(metagenomeSeq)
  design <- model.matrix(~group)
  pheno <- AnnotatedDataFrame(data.frame(group=group))
  rownames(pheno) <- colnames(counts)
  p <- cumNormStatFast(counts)
  dat <- newMRexperiment(counts=counts, phenoData=pheno, featureData = NULL, libSize = colSums(counts), normFactors = metagenomeSeq::calcNormFactors(counts, p=p))
  fit <- fitZig(dat,design)
  lfc <- fit$eb$coefficients[,"group1"]
  pval <- fit$eb$p.value[,"group1"]
  padj <- p.adjust(pval)
  out <- cbind(pval,padj,lfc)
  return(out)
}
```

```{r seurat}
Seurat <- function(counts, group, xlim=NULL, ylim=NULL){
    require(Seurat)
    seur = CreateSeuratObject(counts, project=paste0("cell",as.character(group)), display.progress = FALSE)
    res <- FindMarkers(seur, ident.1 = "cell0", ident.2 = "cell1", print.bar=FALSE)
    pval = res$p_val[match(rownames(counts),rownames(res))]
    padj = p.adjust(pval,"BH")
    out=cbind(pval,padj)
    return(out)
}
```

```{r DEsingle}
#https://github.com/miaozhun/DEsingle downloaded on November 20, 2017
DEsingle <- function(counts, group, xlim=NULL, ylim=NULL){
  require(DEsingle)
  res = DEsignle(counts=counts, group=group)
  pval = res[,"pvalue"]
  padj = res[,"pvalue.adj.FDR"]
  out=cbind(pval,padj)
  return(out)
}
```


## zinbwave

We compute the same weights as zingeR (i.e. posterior probabilities that a count belongs to the count component given that the count and library size is observed), but using the ZINB-WaVE estimation procedure. See more details here (http://www.biorxiv.org/content/early/2017/04/06/125112).

```{r}
NCORES <- 2
registerDoParallel(NCORES)
register(DoparParam())
```

### zinbwave-edgeR
```{r}
zinbwave_edgeR <- function(counts, group, zinb, ylim = NULL, xlim = NULL, main = 'ZINB-WaVE'){
  d=DGEList(counts)
  d=suppressWarnings(edgeR::calcNormFactors(d))
  design=model.matrix(~group)
  weights <- computeObservationalWeights(zinb, counts)
  weights[weights==0] = 1e-15 #errors with zero weights
  d$weights <- weights
  d=estimateDisp(d, design)
  plotBCV(d, ylim = ylim, main = main)
  fit=glmFit(d,design)
  lrt=glmWeightedF(fit,coef=2, independentFiltering = TRUE)
  cbind(pval = lrt$table$PValue, padj =lrt$table$padjFilter)
}
```

### zinbwave-DESeq2
```{r}
zinbwave_DESeq2 <- function(counts, group, zinb){
  colData=data.frame(group=group)
  design=model.matrix(~group)
  dse=DESeqDataSetFromMatrix(countData=counts, colData=colData, design=~group)
  weights <- computeObservationalWeights(zinb, counts)
  weights[weights<1e-6] = 1e-6
  assays(dse)[["weights"]]=weights
  dse = estimateSizeFactors(dse, type="poscounts")
  dse = estimateDispersions(dse, minmu=1e-3)
  #dse = nbinomWaldTest(dse, betaPrior=TRUE, useT=TRUE, df=rowSums(weights)-2, minmu=1e-3)
  dse = nbinomLRT(dse, minmu=1e-3, reduced=~1)
  res = results(dse)
  cbind(pval = res$pvalue, padj = res$padj)
}
```

### zinbwave-limma-voom
```{r}
zinbwave_limma <- function(counts, group, zinb){
  design <- model.matrix(~group)
  nf <- edgeR::calcNormFactors(counts)
  zeroWeights <- computeObservationalWeights(zinb, counts)
  y <- voom(counts, design, plot=FALSE, lib.size = colSums(counts)*nf,
            weights = zeroWeights)
  y$weights <- y$weights * zeroWeights
  fit <- lmFit(y, design, weights=y$weights)
  fit$df.residual <- rowSums(zeroWeights) - ncol(design)
  fit <- eBayes(fit)
  tt <- topTable(fit,coef=2,n=nrow(counts), sort.by = "none")
  pval <- tt$P.Value
  baseMean = unname(rowMeans(sweep(counts,2,nf,FUN="*")))
  hlp <- pvalueAdjustment_kvdb(baseMean=baseMean, pValue=pval)
  padj <- hlp$padj
  cbind(pval = pval, padj = padj)
}
```

# Results

```{r core}
core <- SummarizedExperiment(simData$counts,
                             colData = data.frame(grp = grp))
```

```{r zinbcommondisp}
zinb_c <- zinbFit(core, X = '~ grp', commondispersion = TRUE, epsilon=1e12)
#save(zinb_c, file = 'zinb-common-disp-10xfc2.rda')
#load('zinb-common-disp-10xfc2.rda')

```

```{r zinbgenewisedisp}
 zinb_g <- zinbFit(core, X = '~ grp', commondispersion = FALSE)
# save(zinb_g, file = 'zinb-genewise-disp-10xfc2.rda')
# load('zinb-genewise-disp-10xfc2.rda')
```

# Compare dispersion estimates
```{r tenxDispFC2, warning=FALSE}
counts = simData$counts
myfct = list(DESeq2 = DESeq2,
             edgeR = edgeR,
             limmavoom = limma,
             MAST = MAST,
             NODES = NODES,
             scde = scde,
             metagenomeSeq = metagenomeSeq)

par(mfrow = c(2,2))
ylim = c(0, 11)
xlim = c(0, 16)
res = lapply(myfct, function(fct){
  fct(counts = counts, group = grp, ylim = ylim, xlim = xlim)
})
res[['ZINB-WaVE_DESeq2_common']] = zinbwave_DESeq2(counts, grp, zinb_c)
res[['ZINB-WaVE_edgeR_common']]  = zinbwave_edgeR(counts, grp, zinb_c, ylim=ylim, main = 'ZINB-WaVE, common dispersion', xlim = xlim)
res[['ZINB-WaVE_limmavoom_common']]  = zinbwave_limma(counts, grp, zinb_c)
res[['ZINB-WaVE_DESeq2_genewise']] = zinbwave_DESeq2(counts, grp, zinb_g)
res[['ZINB-WaVE_edgeR_genewise']]  = zinbwave_edgeR(counts, grp, zinb_g, ylim=ylim, main = 'ZINB-WaVE, genewise dispersion', xlim = xlim)
res[['ZINB-WaVE_limmavoom_genewise']]  = zinbwave_limma(counts, grp, zinb_g)
par(mfrow = c(1,1))

#save(res,file="/Users/koenvandenberge/Dropbox/phdKoen/singleCell/zinbwaveZingerGithub/zinbwaveZinger/zinbwaveSimulations/tenX_sims_fc2/res10x.rda")
```

```{r}
res = lapply(res, as.data.frame)
for(k in 1:length(res)) res[[k]]$padj[is.na(res[[k]]$padj)] = 1
```

## Compare weights estimates

The gene-wise dispersion weights seem to be performing less good as compared to the common dispersion weights.

```{r zinbwaveW}
#zingeREdgeRWeights = zingeR::zeroWeightsLS(counts=simData$counts, design=model.matrix(~grp), maxit=250, llTol=1e-1, llOffset=1e-2, verbose=TRUE, plotW=TRUE)
par(mfrow=c(1,2))
zinbwave_c_weights <- computeZinbwaveWeights(zinb_c, counts)
hist(zinbwave_c_weights[simData$dropout==0], main="ZINB-WaVE common: post prob for dropout", breaks=seq(0,1,0.05), cex.main=2/3)
hist(zinbwave_c_weights[simData$dropout==1], main="ZINB-WaVE common: post prob for NB zero", breaks=seq(0,1,0.05), cex.main=2/3)


zinbwave_g_weights <- computeZinbwaveWeights(zinb_g, counts)
 hist(zinbwave_g_weights[simData$dropout==0], main="ZINB-WaVE gene: post prob for dropout", breaks=seq(0,1,0.05), cex.main=2/3)
 hist(zinbwave_g_weights[simData$dropout==1], main="ZINB-WaVE gene: post prob for NB zero", breaks=seq(0,1,0.05), cex.main=2/3)
```


```{r qqplotFC2}
qqplot(zinbwave_c_weights, zinbwave_g_weights, type = 'o',
       main = '',
       xlab = 'ZINB-WaVE weights, common dispersion',
       ylab = 'ZINB-WaVE weights, genewise dispersion')
abline(a=0,b=1)
```

## nDE, TPR, FDR (adjusted pvalue = 0.05)
```{r tenxTableFC2, results = 'asis'}
listRates = lapply(res, function(y){
  nDE = sum(y$padj <= 0.05, na.rm = TRUE)
  TPR = mean(simData$indDE %in% which( y$padj <= 0.05))
  FPR = mean(which(y$padj <= 0.05) %in% simData$indNonDE)
  c(nDE = nDE, TPR = TPR, FPR = FPR)
})

df = do.call(rbind, listRates)
df = as.data.frame(df)
df$Method = names(res)
df$nDE = as.integer(df$nDE)
df$TPR = round(df$TPR*100, 1)
df$FPR = round(df$FPR*100, 1)
df = df[,c('Method', 'nDE', 'TPR', 'FPR')]
colnames(df) = c('Method', 'nDE', 'TPR(%)', 'FPR(%)')
rownames(df) = NULL
kable(df)
```

## TPR vs FDR
```{r truth}
trueDE = rep(0, nTags)
trueDE[simData$indDE] = 1
```



```{r tenxROCfc2}
#all methods
pvalDf = as.data.frame(do.call(cbind, lapply(res, '[[', 1)))
padjDf = as.data.frame(do.call(cbind, lapply(res, '[[', 2)))
pp = COBRAData(pval = pvalDf,
               padj = padjDf,
                truth = data.frame(status = trueDE))
cobraperf <- calculate_performance(pp, binary_truth = "status", thrs = 0.05)
colors=c(limmavoom="blue", "ZINB-WaVE_limmavoom_common"="steelblue", "ZINB-WaVE_limmavoom_genewise"="darkslategray3", edgeR="red", "ZINB-WaVE_edgeR_common"="salmon", "ZINB-WaVE_edgeR_genewise"="deeppink2",  DESeq2="brown",  "ZINB-WaVE_DESeq2_common"="darkseagreen", "ZINB-WaVE_DESeq2_genewise"="darkkhaki",  MAST="darkturquoise", metagenomeSeq="forestgreen", scde="grey", NODES="black",  Seurat="dodgerblue")
#iCOBRA converts '-' to '.'. Redo this.
cobraNames = sort(names(cobraperf@overlap)[1:(ncol(cobraperf@overlap)-1)])
cobraNames = gsub(x=cobraNames, pattern=".", fixed=TRUE, replacement="-")
colsCobra=colors[match(cobraNames,names(colors))]
cobraplot <- prepare_data_for_plot(cobraperf, colorscheme=colsCobra)
#save(cobraplot,file="cobraPlot10x.rda")
png("~/Dropbox/phdKoen/singleCell/zinbwaveZingeR/plots2/scSimulation_10x_allMethods.png", width=7,height=8, units="in", res=300)
plot_fdrtprcurve(cobraplot, pointsize=2) + xlab("FDP")
dev.off()

# no genewise zinbwave
pvalDfCommon = pvalDf[,-grep(x=colnames(pvalDf), pattern="genewise")]
padjDfCommon = padjDf[,-grep(x=colnames(padjDf), pattern="genewise")]
pp = COBRAData(pval = pvalDfCommon,
               padj = padjDfCommon,
                truth = data.frame(status = trueDE))
cobraperf <- calculate_performance(pp, binary_truth = "status", thrs = 0.05)
colors=c(limmavoom="blue", "ZINB-WaVE_limmavoom_common"="steelblue", "ZINB-WaVE_limmavoom_genewise"="darkslategray3", edgeR="red", "ZINB-WaVE_edgeR_common"="salmon", "ZINB-WaVE_edgeR_genewise"="deeppink2",  DESeq2="brown",  "ZINB-WaVE_DESeq2_common"="darkseagreen", "ZINB-WaVE_DESeq2_genewise"="darkkhaki",  MAST="darkturquoise", metagenomeSeq="forestgreen", scde="grey", NODES="black",  Seurat="dodgerblue")
#iCOBRA converts '-' to '.'. Redo this.
cobraNames = sort(names(cobraperf@overlap)[1:(ncol(cobraperf@overlap)-1)])
cobraNames = gsub(x=cobraNames, pattern=".", fixed=TRUE, replacement="-")
colsCobra=colors[match(cobraNames,names(colors))]
cobraplot <- prepare_data_for_plot(cobraperf, colorscheme=colsCobra)
#save(cobraplot,file="cobraplot10xNoGenewise.rda")
plot_fdrtprcurve(cobraplot, pointsize=2) + xlab("FDP")

# no genewise zinbwave, ZINB-WaVE limma
pvalDf2 = pvalDfCommon[,-grep(x=colnames(pvalDfCommon), pattern="ZINB-WaVE_limmavoom")]
padjDf2 = padjDfCommon[,-grep(x=colnames(padjDfCommon), pattern="ZINB-WaVE_limmavoom")]
pp = COBRAData(pval = pvalDf2,
               padj = padjDf2,
                truth = data.frame(status = trueDE))
cobraperf <- calculate_performance(pp, binary_truth = "status", thrs = 0.05)
colors=c(limmavoom="blue", "ZINB-WaVE_limmavoom_common"="steelblue", "ZINB-WaVE_limmavoom_genewise"="darkslategray3", edgeR="red", "ZINB-WaVE_edgeR_common"="salmon", "ZINB-WaVE_edgeR_genewise"="deeppink2",  DESeq2="brown",  "ZINB-WaVE_DESeq2_common"="darkseagreen", "ZINB-WaVE_DESeq2_genewise"="darkkhaki",  MAST="darkturquoise", metagenomeSeq="forestgreen", scde="grey", NODES="black",  Seurat="dodgerblue")
#iCOBRA converts '-' to '.'. Redo this.
cobraNames = sort(names(cobraperf@overlap)[1:(ncol(cobraperf@overlap)-1)])
cobraNames = gsub(x=cobraNames, pattern=".", fixed=TRUE, replacement="-")
colsCobra=colors[match(cobraNames,names(colors))]
cobraplot <- prepare_data_for_plot(cobraperf, colorscheme=colsCobra)
save(cobraplot,file="cobraPlot10xNoLimma.rda")
png("~/Dropbox/phdKoen/singleCell/zinbwaveZinger/plots2/scSimulation_10x_cutoff_noLimma.png", width=7,height=8, units="in", res=300)
plot_fdrtprcurve(cobraplot, pointsize=2, xaxisrange=c(0,0.41), yaxisrange=c(0,0.4)) + xlab("FDP")
dev.off()


#orderLegend = c(2, 9, 6, 1, 8, 5, 3, 10, 7, 4)
# p2 <- plot_fdrtprcurve(cobraplot, plottype = c("curve", "points"), pointsize = .2,
#                        linewidth = .5, xaxisrange = c(0, .5)) +
#   theme(legend.text=element_text(size=7)) +
#   scale_color_manual(labels = names(colsCobra)[orderLegend],
#                      values = unname(colsCobra)[orderLegend],
#                      name = 'Method')
# legend <- get_legend(p2)
#
# plot_grid(p1, legend, nrow = 1, ncol = 2, rel_widths = c(1, .4))

```

## Distribution of pvalues

```{r tenxPvaluesFC2zinbwave}
ylim = c(0, 3000)
par(mfrow = c(2,3))
hist = lapply(c(8:13), function(i){
  hist(res[[i]][,'pval'], main = names(res)[i], ylim = ylim, xlab = 'pvalues')
})
par(mfrow = c(1,1))
```

```{r tenxPvaluesFC2}
ylim = c(0, 3000)
par(mfrow = c(3,3))
hist = lapply(c(1:3, 5:10), function(i){
  main = gsub('_common', '', names(res)[i])
  hist(res[[i]][,'pval'], main = main, ylim = ylim, xlab = 'pvalues')
})
par(mfrow = c(1,1))
```

