---
title: "Evaluate FPR on 10x Genomics PBMC dataset"
author: "Fanny Perraudeau"
date: "`r Sys.Date()`"
output: 
  html_document: 
    fig_height: 7
    fig_width: 7
    toc: yes
    code_folding: hide
    toc_float: yes
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE, include=FALSE, autodep=TRUE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(DESeq2)
library(edgeR)
library(limma)
library(zinbwave)
library(BiocParallel)
library(doParallel)
library(Biobase)
library(ggplot2)
```

```{r}
NCORES <- 2
registerDoParallel(NCORES)
register(DoparParam())
```

The goal of this document is to evaluate the FPR of our methods and compare it to the FPR of other methods (edgeR, DESeq2, limmavoom, MAST, scde) on 10x-genomics data 3k PMBCs.

# Functions

```{r sourceEdgeR}
runEdgeR <- function(e) {
  library(edgeR)
  condition = pData(e)$condition
  design <- model.matrix(~ condition)
  dgel <- DGEList(exprs(e))
  dgel <- edgeR::calcNormFactors(dgel)
  dgel=estimateDisp(dgel,design)
  edger.fit <- glmFit(dgel, design)
  edger.lrt <- glmLRT(edger.fit, coef="conditionB")
  pvals <- edger.lrt$table$PValue
  padj <- p.adjust(pvals,method="BH")
  padj[is.na(padj)] <- 1
  logfc <- edger.lrt$table$logFC
  list(pvals=pvals, padj=padj, logfc = logfc)
}

```

```{r deseq2}
runDESeq2_poscounts <- function(e) {
  library(DESeq2)
  dds <- DESeqDataSetFromMatrix(exprs(e),
                                colData=DataFrame(pData(e)),
                                design=~ condition)
  dds <- estimateSizeFactors(dds,type="poscounts")
  dds <- estimateDispersions(dds, minmu = 1e-3)
  #dds <- nbinomWaldTest(dds, betaPrior=TRUE, modelMatrixType="standard", minmu = 1e-3)
  dds <- nbinomLRT(dds, minmu=1e-3, reduced=~1)
  res <- results(dds, name="condition_B_vs_A")
  pvals <- res$pvalue
  padj <- res$padj
  logfc <- res$log2FoldChange
  list(pvals=pvals, padj=padj, logfc = logfc)
}
```

```{r zinbwaveEdgeR}
runZinbwaveEdgeR <- function(e){
  condition = pData(e)$condition
  design <- model.matrix(~ condition)
  # compute zinbwave weights
  library(zinbwave)
  zinb <- zinbFit(exprs(e), X = design, epsilon = 1e12)
  weights <- computeObservationalWeights(zinb, exprs(e))
  # use -edgeR
  d <- DGEList(exprs(e))
  d <- edgeR::calcNormFactors(d)
  d$weights <- weights
  d=estimateDisp(d, design)
  fit=glmFit(d,design)
  lrt=glmWeightedF(fit,coef=2, independentFiltering = TRUE)
  pvals = lrt$table$PValue
  list(pvals = pvals, padj = lrt$table$padjFilter, logfc = lrt$table$logFC)
}
```

```{r zinbwavedeseq2}
runZinbwaveDESeq2 <- function(e){
  condition = pData(e)$condition
  design <- model.matrix(~ condition)
  # compute zinbwave weights
  library(zinbwave)
  zinb <- zinbFit(exprs(e), X = design, epsilon = 1e12)
  weights <- computeObservationalWeights(zinb, exprs(e))
  # use -DESeq2
  dse <- DESeqDataSetFromMatrix(exprs(e),
                                colData=DataFrame(pData(e)),
                                design=~ condition)
  dse <- estimateSizeFactors(dse, type="poscounts")
  dimnames(weights) = NULL
  assays(dse)[["weights"]] = weights
  dse <- estimateDispersions(dse, minmu = 1e-3)
  #dse = nbinomWaldTest(dse, betaPrior=TRUE, useT=TRUE, df=rowSums(weights)-2, minmu = 1e-3)
  dse <- nbinomLRT(dse, minmu=1e-3, reduced=~1)
  res = results(dse, name="condition_B_vs_A")
  list(pvals = res$pvalue, padj = res$padj, logfc = res$log2FoldChange)
}
```

```{r}
runMAST <- function(e){
  require(MAST)
  counts <- exprs(e)
  tpm <- counts*1e6/colSums(counts)
  tpm <- log2(tpm+1)
  sca <- FromMatrix(tpm,cData=data.frame(group=pData(e)$condition))
  
  # Adaptive thresholding from MAST vignette for MAIT analysis
  freq_expressed <- 0.05
  thres <- thresholdSCRNACountMatrix(assay(sca), nbins = 300,
                                     conditions = pData(e)$condition)
  #par(mfrow=c(5,4))
  #plot(thres)
  assays(sca) <- list(thresh=thres$counts_threshold, tpm=assay(sca))
  expressed_genes <- freq(sca) > freq_expressed
  sca <- sca[expressed_genes,]

  ngeneson <- apply(exprs(e),2,function(x) mean(x>0))
  CD <- colData(sca)
  CD$ngeneson <- ngeneson
  CD$cngeneson <- CD$ngeneson-mean(ngeneson)
  colData(sca) <- CD
  ## differential expression
  fit <- zlm(~ cngeneson + group, sca = sca,
             method = "bayesglm", ebayes = TRUE)
  L <- matrix(0, nrow = ncol(coef(fit, "D")))
  rownames(L) <- colnames(coef(fit, "D"))
  L["groupB",] <- 1
  lrFit <- lrTest(fit, hypothesis = L)
  pval <- lrFit[, 'hurdle', 'Pr(>Chisq)']
  padj <- p.adjust(pval, method = "BH")
  list(pvals = pval, padj = padj, logfc = NA)
}
```

```{r limmavoom}
runVoom <- function(e) {
  library(limma)
  condition = pData(e)$condition
  design <- model.matrix(~ condition)
  dgel <- DGEList(exprs(e))
  dgel <- edgeR::calcNormFactors(dgel)
  v <- voom(dgel,design,plot=FALSE)
  fit <- lmFit(v,design)
  fit <- eBayes(fit)
  tt <- topTable(fit,coef="conditionB",n=nrow(dgel),sort.by="none")
  pvals <- tt$P.Value
  padj <- p.adjust(pvals,method="BH")
  padj[is.na(padj)] <- 1
  list(pvals=pvals, padj=padj, logfc=tt$logFC)
}
```

```{r scde}
runSCDE <- function(e){
  require(scde)
  # calculate models
  counts = exprs(e)
  counts<-apply(counts,2,function(x) {storage.mode(x) <- 'integer'; x}) 
  rownames(counts) = fData(e)[,1]
  o.ifm <- scde.error.models(counts = counts, groups = pData(e)$condition, n.cores = 2,
                             threshold.segmentation = TRUE,
                             save.crossfit.plots = FALSE, 
                             save.model.plots = FALSE, verbose = 0)
  
  # estimate gene expression prior
  o.prior <- scde.expression.prior(models = o.ifm, counts = exprs(e), length.out = 400, show.plot = FALSE)
  # run differential expression tests on all genes.
  ediff <- scde.expression.difference(o.ifm, exprs(e), o.prior, 
                                      groups  =  pData(e)$condition,
                                      n.randomizations  =  100,
                                      n.cores  =  2, verbose  =  0,
                                    batch=factor(pData(e)$pickingSession))
  pvals=(1-pnorm(abs(ediff$Z)))*2
  padj=p.adjust(pvals,method="BH")
  list(pvals=pvals,padj=padj,logfc=NA)
}
```


# Data

```{r data}
load('../../realdata/data/core.rda')
core
ncells = 45
nreps = 30

subsets = sapply(1:nreps, function(i){
  set.seed(i)
  c1 = sample(ncol(core), ncells, replace = FALSE)
  remaining = seq_len(ncol(core))
  remaining = remaining[!remaining %in% c1]
  c2 = sample(remaining, ncells, replace = FALSE)
  stopifnot(sum(c1 %in% c2) == 0 )
  stopifnot(sum(c2 %in% c1) == 0 )
  c(c1, c2)
})
subsets = as.data.frame(t(subsets))
```

# Results

```{r res}
algos <- list("DESeq2"          = runDESeq2_poscounts,
              "DESeq2-zinbwave" = runZinbwaveDESeq2,
              "edgeR"           = runEdgeR,
              "edgeR-zinbwave"  = runZinbwaveEdgeR,
              "limma-voom"      = runVoom,
              "MAST"            = runMAST,
              "SCDE"            = runSCDE)
namesAlgos <- names(algos)
names(namesAlgos) <- namesAlgos
```

```{r,eval=FALSE}
eset = as(core, 'ExpressionSet')
res <- lapply(1:nreps, function(i) {
    cat(i," ")
    eLoop <- eset[,as.numeric(subsets[i,])]
    ngenes = sum(rowSums(exprs(eLoop) > 0) < 5)
    cat(sprintf("Removing %s genes with only 0's", ngenes))
    # remove genes with too many zeros
    eLoop <- eLoop[rowSums(exprs(eLoop) > 0) > 4, ]
    condition = factor(rep(c("A", "B"), each = 45))
    pData(eLoop)$condition = condition
    resFPR <- lapply(namesAlgos, function(n){
      print(n)
      algos[[n]](eLoop)
    })
    save(resFPR, file = sprintf('FPR_%s.rda', i))
    rm(resFPR)
    i
})
```

```{r}
res <- lapply(1:nreps, function(i) {
  load(sprintf('FPR_%s.rda', i))
  resFPR
})
```

## Boxplots
```{r prepBoxplots}
hlp=lapply(res,function(replication){
  lapply(replication,function(method){
    pval=method$pvals
    pval[is.na(pval)]=1 #independent filtering
    mean(pval<=0.05)
  })
})

fprHat=Reduce(hlp,f=cbind)
fprHat=matrix(unlist(fprHat),nrow=length(algos),ncol=nreps,byrow=FALSE)
rownames(fprHat)=namesAlgos

boxplotData=data.frame(fpr=c(t(fprHat)),method=rep(namesAlgos,each=nreps))
```

```{r fpr10xPCER}
head(boxplotData,2)
boxplotData$method = as.vector(boxplotData$method) 
ourcol = rep(1, length(unique(boxplotData$method)))
ourcol[c(1, 4)] = 4 #our methods in blue
ggplot(boxplotData,
       aes(x=reorder(method,fpr,median),y=fpr)) +
  geom_boxplot(outlier.colour=rgb(0,0,0,0), col = ourcol) +
  theme_bw() +
    geom_point(position = position_jitter(w = 0.1, h = 0),
               color="grey50", size=1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + xlab("") +
  scale_colour_discrete(guide="none") + ylab("PCER") +
  geom_hline(aes(yintercept=0.05,colour="red")) +
  theme(text = element_text(size=20))
```

## Histograms of pvalues
```{r fpr10xPval}
library(scales)
par(mfrow=c(3,3))
hh <- sapply(1:length(res[[1]]), function(i){
  col = 'white'
  if (i %in% c(2,4)) col = 'blue'
  nn = names(algos)[i]
  hist(res[[1]][[i]]$pvals, main = nn,
       xlab = "p-values", cex.lab = 1.5, 
       cex.axis =  1.5, cex.main = 1.5, col = alpha(col, .8),
       breaks = seq(0, 1, by = 0.05)) 
})
par(mfrow=c(1,1))
```

## Plot paper

```{r fpr10x,fig.width=10,fig.height=5}
par(mar = c(4.1, 4.25, 3, 1), bty = "l")
layout(cbind(matrix(1,3,3), matrix(2:10,3,3,byrow = TRUE)))
# pcer = f(eps)
library(gridBase)
library(grid)
plot.new()              ## suggested by @Josh
vps <- baseViewports()
pushViewport(vps$figure) ##   I am in the space of the autocorrelation plot
vp1 <-plotViewport(c(1, 1, 1, 1)) ## create new vp with margins, you play with this values 

h <- ggplot(boxplotData,
       aes(x=reorder(method,fpr,median),y=fpr)) +
  geom_boxplot(outlier.colour=rgb(0,0,0,0), col = ourcol) +
  theme_bw() +
    geom_point(position = position_jitter(w = 0.1, h = 0),
               color="grey50", size=1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + xlab("") +
  scale_colour_discrete(guide="none") + ylab("PCER") +
  geom_hline(aes(yintercept=0.05,colour="red")) +
  theme(text = element_text(size=20))

print(h,vp = vp1)  
mtext("a" ,at=-0.15, padj = -1, font=2, cex=4/3)

# hist pvalues
hh <- sapply(1:length(res[[1]]), function(i){
  col = 'white'
  if (i %in% c(2,4)) col = 'blue'
  nn = names(algos)[i]
  hist(res[[1]][[i]]$pvals, main = nn,
       xlab = "p-values", cex.lab = 1.5, 
       cex.axis =  1.5, cex.main = 1.2, col = alpha(col, .8),
       breaks = seq(0, 1, by = 0.05)) 
  if (i==2) mtext("b" ,at=-0.7, padj = -1, font=2, cex=4/3)
})
```


# sessionInfo
```{r}
sessionInfo()
```